File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/ChakraCrystalItem.java
package de.limbusdev.guardianmonsters.guardians.items;

import de.limbusdev.guardianmonsters.guardians.Element;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;

public class ChakraCrystalItem extends Item
{
    private Element element;

    public ChakraCrystalItem(String name, String element)
    {
        super(name, Category.CHAKRACRYSTAL);
        try {
            this.element = Element.valueOf(element.toUpperCase());
        } catch (Exception e) {
            e.printStackTrace();
            this.element = Element.NONE;
        }
    }

    /**
     * calculates the chance to ban a guardian with this item
     * @param guardian
     * @return
     */
    public float chance(AGuardian guardian)
    {
        float chance = 0f;
        float elementFactor;

        if(guardian.getSpeciesDescription().getElements(
                guardian.getAbilityGraph().getCurrentForm()
        ).contains(element, false)) {
            elementFactor = 2f;
        } else {
            elementFactor = 1f;
        }

        chance = (1f - guardian.getIndividualStatistics().getHpFraction()/elementFactor/100f/2f - guardian.getIndividualStatistics().getLevel()/100f);

        if(chance > 1f) {
            chance = 1f;
        }

        return chance;
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/ChakraCrystalItem.kt
package de.limbusdev.guardianmonsters.guardians.items

import de.limbusdev.guardianmonsters.guardians.Element
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.utils.extensions.f

class ChakraCrystalItem(name: String, element: String) : Item(name, Category.CHAKRACRYSTAL)
{
    // .................................................................................. Properties
    private var element: Element = Element.NONE

    // ................................................................................ Constructors
    init
    {
        try
        {
            this.element = Element.valueOf(element.toUpperCase())
        }
        catch (e: Exception)
        {
            e.printStackTrace()
            this.element = Element.NONE
        }
    }

    // ..................................................................................... Methods
    /**
     * calculates the chance to ban a guardian with this item
     * @param guardian
     * @return
     */
    fun chance(guardian: AGuardian): Float
    {
        var chance = 0f
        val elementFactor: Float = if (guardian.isOfElement(element)) { 2f }
                                   else                               { 1f }

        chance =  1f - guardian.stats.hpFraction
        chance /= elementFactor / 100f / 2f
        chance -= guardian.stats.level / 100f

        if (chance > 1f) { chance = 1f }

        return chance
    }
}
